### PR TITLE
Wandb

### PR DESCRIPTION
**Pull request by MatteoKartoon**
* weight and biases implementation
* move `bce_with_logits` option to config
* aggregate loss across GPUs at the epoch-level
* split the `train_epoch` function into smaller functions

### CODE REVIEW

File: scripts/train.py [hunk 1]
```diff
@@ -271,11 +299,26 @@ def train_epoch(self, epoch):
...
+        epoch_losses_train.append(self.loss_log.avg)
+        # After epoch
+        # Stack and gather losses from all processes
+        epoch_losses_tensor = torch.tensor(epoch_losses_train, device=accelerator.device)
+        all_losses = accelerator.gather(epoch_losses_tensor)
+
+        # Compute average loss on main process
+        if accelerator.is_main_process:
+            avg_loss = all_losses.mean().item()
+            logger.info('@==Final== Epoch[{0}/{1}]  Average Training Loss: {loss:.3f}'.format(epoch, args.epochs, loss=avg_loss))
```
Comment (joelkartoon) [inline scripts/train.py:313]: nit: same

File: scripts/train.py [hunk 2]
```diff
@@ -242,6 +267,8 @@ def _batch(self, batch, batch_idx, epoch_num, validation=False):
                     total_norm += param_norm.item() ** 2
                 total_norm = total_norm ** 0.5
                 print("Gradient norm:", total_norm)
+                if accelerator.is_main_process:
+                    wandb.log({"Gradient norm": total_norm})
```
Comment (joelkartoon) [inline scripts/train.py:271]: for all the `wandb.log`, give the `step` parameter

File: scripts/train.py [hunk 3]
```diff
@@ -286,9 +329,21 @@ def train_epoch(self, epoch):
                 for loss_name, loss_value in self.loss_dict_validation.items():
                     info_loss += ', {}: {:.3f}'.format(loss_name, loss_value)
                 logger.info(' '.join((info_progress, info_loss)))
-        info_loss = '@==Final== Epoch[{0}/{1}]  Validation Loss: {loss.avg:.3f}'.format(epoch, args.epochs, loss=self.val_loss_log)
+        info_loss = '@==Final== Epoch[{0}/{1}]  Validation Loss Device {2}: {loss.avg:.3f}'.format(epoch, args.epochs, accelerator.device, loss=self.val_loss_log)
```
Comment (joelkartoon) [inline scripts/train.py:332]: factorize this into a function that is re-used across training & validation

File: scripts/train.py [hunk 4]
```diff
@@ -50,7 +51,31 @@
 config = Config()
 if config.rand_seed:
     set_seed(config.rand_seed)
-    
+
+if accelerator.is_main_process:
+    wandb.init(
```
Comment (joelkartoon) [inline scripts/train.py:56]: move this whole block of code into a util function called `_init_wandb`

File: scripts/train.py [hunk 5]
```diff
@@ -209,6 +188,42 @@ def __init__(
         self.loss_log = AverageMeter()
         self.val_loss_log = AverageMeter()
 
+        self.step_counter=0
+
+    def iteration_over_batches(self, epoch,loader, dict_losses, log_losses, log_each_steps, validation=False, log_task="Training"):
```
Comment (joelkartoon) [inline scripts/train.py:193]: if the only purpose of `log_task` is to be `training` or `validation` you could do
log_task = "validation" if validation else "training"

File: scripts/train.py [hunk 6]
```diff
@@ -188,39 +188,77 @@ def __init__(
         self.loss_log = AverageMeter()
         self.val_loss_log = AverageMeter()
 
-        self.step_counter=0
+    def iteration_over_batches_validation(self, epoch,loader, dict_losses, info_progress=None, step_idx=None, training_result=None):
+        #Initialize a dictionary containing the average of the losses
+        dict_losses_average={k:0 for k in dict_losses.keys()}
```
Comment (joelkartoon) [inline scripts/train.py:193]:
```
total_loss = 0
for batch_idx [...]:
  total_loss += dict_losses["loss_pix"]
overall_avg_loss = total_loss / len(xxx)`
```

File: scripts/train.py [hunk 7]
```diff
@@ -188,39 +188,77 @@ def __init__(
         self.loss_log = AverageMeter()
         self.val_loss_log = AverageMeter()
 
-        self.step_counter=0
+    def iteration_over_batches_validation(self, epoch,loader, dict_losses, info_progress=None, step_idx=None, training_result=None):
```
Comment (joelkartoon) [inline scripts/train.py:191]: replace `dict_losses`  -> `self.validation_loss_dict`
`loader` -> `self.[..]`
in the function, and remove from args

File: scripts/train.py [hunk 8]
```diff
@@ -188,39 +188,77 @@ def __init__(
...
+            self._batch(batch, batch_idx, epoch, validation=True)
+            # Logger
+            for loss_name, loss_value in dict_losses.items():
+                dict_losses_average[loss_name]+=loss_value
+        #Compute the average of the losses over the validation set
+        for loss_name, loss_value in dict_losses_average.items():
+            dict_losses_average[loss_name]/=len(loader)
+        #For each loss type, compute the average between the devices
+        info_loss = f'Validation Losses'
+        for loss_name, loss_value in dict_losses_average.items():
```
Comment (joelkartoon) [inline scripts/train.py:205]: roughly something like this
```
loss_general_value=self.average_between_devices(loss_value)
accelerator.wait_for_everyone()
logger.info(f"Validation Loss: {overall_avg_loss}")
if accelerator.is_main_process:
            wandb.log({"Validation Loss": loss_general_value_pixel, "Training Loss": training_result},step=step_idx)
```

File: scripts/train.py [hunk 9]
```diff
@@ -188,39 +188,77 @@ def __init__(
...
-            # with nullcontext if not args.use_accelerate or accelerator.gradient_accumulation_steps <= 1 else accelerator.accumulate(self.model):
-            self._batch(batch, batch_idx, epoch, validation)
+            step_idx=batch_idx+len(loader)*(epoch-config.start_epoch)
+            self._batch(batch, batch_idx, epoch)
             # Logger
             if batch_idx % log_each_steps == 0:
                 info_progress = f'Epoch[{epoch}/{args.epochs}] Iter[{batch_idx}/{len(loader)}].'
-                info_loss = f'{log_task} Losses'
+                info_loss = f'Training Losses'
                 for loss_name, loss_value in dict_losses.items():
```
Comment (joelkartoon) [inline scripts/train.py:227]: simplify in similar way

File: scripts/train.py [hunk 10]
```diff
@@ -188,39 +188,77 @@ def __init__(
...
+                    loss_general_value=self.average_between_devices(loss_value)
+                    accelerator.wait_for_everyone()
+                    loss_general_value_pixel=None
+                    if loss_general_value is not None:
+                        if loss_name=="loss_pix":
+                            loss_general_value_pixel=loss_general_value
+                        #add to the print string
+                        info_loss += f', {loss_name}: {loss_general_value}'
+                        logger.info(' '.join((info_progress, info_loss)))
+                self.loss_dict_validation = {k:0 for k in self.loss_dict_train.keys()}
```
Comment (joelkartoon) [inline scripts/train.py:238]: remove as part of simplification

File: scripts/train.py [hunk 11]
```diff
@@ -188,39 +188,77 @@ def __init__(
...
+                self.iteration_over_batches_validation(epoch,self.validation_loader, self.loss_dict_validation, info_progress, step_idx, loss_general_value_pixel)
+
+    def average_between_devices(self, values):
```
Comment (joelkartoon) [inline scripts/train.py:241]: nit: use typing

`values: List[float]`

File: scripts/train.py [hunk 12]
```diff
@@ -188,39 +188,77 @@ def __init__(
...
+        devices_values=[]
+        devices_values.append(values)
+        #gather values from all processes
+        values_tensor = torch.tensor(devices_values, device=accelerator.device)
```
Comment (joelkartoon) [inline scripts/train.py:245]:
```
values_tensor = torch.tensor([values], device=accelerator.device)
```

File: scripts/train.py [hunk 13]
```diff
@@ -188,39 +188,77 @@ def __init__(
...
+        all_values = accelerator.gather(values_tensor)
+
+        # Compute average loss on main process
+        if accelerator.is_main_process:
+            avg_loss = all_values.mean().item()
```
Comment (joelkartoon) [inline scripts/train.py:250]:
```
return all_values.mean().item()
```

File: scripts/train.py [hunk 14]
```diff
@@ -188,39 +188,77 @@ def __init__(
...
-        epoch_losses.append(log_losses.avg)
-        #gather losses from all processes
-        epoch_losses_tensor = torch.tensor(epoch_losses, device=accelerator.device)
-        all_losses = accelerator.gather(epoch_losses_tensor)
-
-        # Compute average loss on main process
-        if accelerator.is_main_process:
-            avg_loss = all_losses.mean().item()
+        avg_loss = self.average_between_devices(log_losses.avg)
+        if avg_loss is not None:
```
Comment (joelkartoon) [inline scripts/train.py:260]: `if accelerator.is_main_process()`

File: scripts/train.py [hunk 15]
```diff
@@ -275,7 +312,7 @@ def _batch(self, batch, batch_idx, epoch_num, validation=False):
             self.optimizer.step()
 
             # Print gradient norm to monitor training
-            if batch_idx % 10 == 0:
+            if batch_idx % 15 == 0:
```
Comment (joelkartoon) [inline scripts/train.py:315]: avoid hard-coded values

File: scripts/train.py [hunk 16]
```diff
@@ -303,14 +348,11 @@ def train_epoch(self, epoch):
             self.loss_log = AverageMeter()
             self.val_loss_log = AverageMeter()
 
-        self.iteration_over_batches(epoch, self.train_loader, self.loss_dict_train, self.loss_log, 10)
-        self.epoch_final_logs(epoch, self.loss_log)
+        self.iteration_over_batches_train(epoch, self.train_loader, self.loss_dict_train, log_each_steps=15)
```
Comment (joelkartoon) [inline scripts/train.py:351]: use same variable

File: scripts/train.py [hunk 17]
```diff
@@ -188,76 +188,55 @@ def __init__(
...
         #Print the final epoch logs
         info_loss = f'@==Final== Epoch[{epoch}/{args.epochs}]  {log_task} Loss Device {accelerator.device}: {log_losses.avg}'
         logger.info(info_loss)
         avg_loss = self.average_between_devices(log_losses.avg)
-        if avg_loss is not None:
+        accelerator.wait_for_everyone()
+        if accelerator.is_main_process:
             logger.info(f'@==Final== Epoch[{epoch}/{args.epochs}]  Average {log_task} Loss: {avg_loss}')
         # Synchronize before next steps
         accelerator.wait_for_everyone()
```
Comment (joelkartoon) [inline scripts/train.py:242]: remove this one

**Review by joelkartoon — APPROVED** [review]
